zip-0308.rst

```diff
@@ -0,0 +1,365 @@
+unique solution and be practically solvable for the number of transactions
+involved. The chosen distribution of transaction amounts mitigates this
+potential vulnerability by ensuring that there will be many solutions for sets
+of transactions, including "incorrect" solutions (that is, solutions that mix
+transactions from different users, contrary to the supposed adversary's
+inference).
+
+Making the chosen amount inclusive of the fee avoids leaving any unmigrated
+funds at the end, in the case where the original amount to migrate was a
+multiple of 0.01 ZEC.
+
+
+Other design decisions
+----------------------
+
+We assume use of the normal wallet note selection algorithm and change
+handling. Change is sent back to the default address, which is the z-address
+of the first selected Sprout note. The number of JoinSplits will therefore be
+the same as for a normal transaction sending the same amount with the same
+wallet state. Only the ``vpub_out`` of the last JoinSplit will be nonzero.
+There will always be exactly one Sapling Output.
+
+The expiration height is chosen as for other transactions.
+
+The fee for each migration transaction MUST be 0.0001 ZEC. This fee is taken
+from the funds to be migrated.
+
+Some wallets by default add a "developer fee" to each transaction, directed
+to the developer(s) of the wallet. This is typically implemented by adding
+the developer address as an explicit output, so if migration transactions are
+generated internally by ``zcashd``, they will not include the developer fee.
+We strongly recommend *not* patching the ``zcashd`` code to add the developer
+fee output to migration transactions, because doing so partitions the anonymity
+set between users of that wallet and other users.
+
+There MUST NOT be any transparent inputs or outputs, or Sapling Spends, in a
+migration transaction.
+
+The ``lock_time`` field MUST be set to 0 (unused).
+
+
+Open questions
+--------------
+
+The above strategy has several "magic number" parameters:
+
+* the interval between batches (500 blocks)
+* the maximum number of transactions in a batch (5)
+* the distribution of exponents (uniform integer in 6..9)
+* the distribution of mantissae (uniform integer in 1..99).
+
+These have been chosen by guesswork. Should we change any of them?
+
+In particular, if the amount to migrate is large, then this strategy can
+result in large amounts (up to 990 ZEC, worth USD ~49500 at time of writing)
+transferred in each transaction. This leaks the fact that the transaction was
+sent by a user who has at least that amount.
+
+The strategy does not migrate any remaining fractional amount less than
+0.01 ZEC (worth USD ~0.50 at time of writing). Is this reasonable?
+
+In deciding the amount to send in each transaction, the strategy does not
+take  account of the values of individual Sprout notes, only the total amount
+remaining to migrate. Can a strategy that is sensitive to individual note
+values improve privacy?
```

Comment:
Given we are only requiring that migration transactions be hidden amongst themselves, what information can we infer from the number of inputs and outputs, knowing that it is part of a migration?

- On the receive side, there will be a single Sapling output for every one of these transactions, so everything will be identical there.
- On the spend side, we will have however many input notes are required to reach the desired target value, and at least one change note. If more than two notes are required to fun the selected amount, then we will have one JoinSplit per input. If the selected amount can be funded with one or two notes, we will have a single JoinSplit.

If the selection strategy is not sensitive to individual note values, then it is possible to end up creating a transaction that consumes most or all of the Sprout notes in the wallet. That combined with the amount revealed might be usable as an indicator that the revealed value is close-to-all of the (remaining) wallet balance. This might particularly affect e.g. wallet developers who received many small notes, which would appear as spikes in the number of inputs. On the other hand, the note-selection strategy in `zcashd` is largest-first, which tends to generate fragmentation (though I do not have any metrics on this).

Separately, there's a potential performance issue, where if the selected value requires many notes to fund, and each note requires a 20-second proof generation, the time to generate a single transaction may exceed the window this ZIP aims for, reducing the indistinguishability of the transaction (either because it appears later, or has to choose an earlier anchor).

So my feeling is that, while we don't necessarily need to be sensitive to the individual note values, we _probably_ should be sensitive to the number of notes required to fill a given random coin. At a minimum, we should probably re-draw the coin if filling it would require an excessive number of notes.

Reply:
The note selection done by `find_unspent_notes` https://github.com/zcash/zcash/blob/87171d35b1c2c99b86edf6327e16967226096c9d/src/wallet/asyncrpcoperation_sendmany.cpp#L1090-L1094 uses largest-note-first.

While technically this does leak some information, it's not clear that taking account of note values would help much (and it could hurt if not done carefully). On balance I think it would overcomplicate the spec and implementation.